### PR TITLE
Make null Variables on convolution and batchnorm work / Use Toffee IR schema to disambiguate things

### PR DESCRIPTION
**Pull request by ezyang**
This addresses when bias is disabled, which occurs in torchvision's
alexnet and densenet.

The general strategy is this:

- When we encounter a null variable, we turn this into a Constant
  node with an undefined at::Tensor

- Toffee exports for BatchNorm and Conv have special cases for bias,
  checking if they are provided by a Constant node with undefined
  value, and just omit the input if so.

----

Also:

Let say I write alpha=2 in my PyTorch code.  Is alpha a float
or an int?  This problem is resolved when we actually pass
it to the underlying kernel, which knows what type it expects
it as.

When serializing to Toffee IR, the Toffee NodeProto also needs
to dictate the correct type; otherwise, we may guess wrong.
We get this information from the OpSchema in the ToffeeIR library.
With this, we can avoid explicitly casting in dropout.py and
auto_primspec.py

TODO: Unclear if this will work on the jit_dirty branch!

### CODE REVIEW

**Comment by zdevito** [comment]
https://github.com/ezyang/pytorch/pull/97 already has code to format attributes. Let's not merge conflict it to death.

File: torch/csrc/autograd/functions/special.cpp [hunk 1]
```diff
@@ -26,8 +26,10 @@ auto Eval::getSubgraph(const variable_list& inputs, const variable_list& outputs
   // Prepare a set of all edges that shouldn't be followed during the search
   edge_set input_edges;
   input_edges.reserve(inputs.size());
-  for (auto & input : inputs)
+  for (auto & input : inputs) {
+    if (!input) continue;
```
Comment (zdevito) [inline torch/csrc/autograd/functions/special.cpp]: This stuff is temporary, right?

File: torch/csrc/jit/attributes.h [hunk 2]
```diff
@@ -53,6 +59,15 @@ struct VectorAttributeValue : public AttributeValue {
     auto copy = value_;
     return Ptr(new VectorAttributeValue(name, std::move(copy)));
   }
+  virtual void printValue(std::ostream & out) const override {
```
Comment (zdevito) [inline torch/csrc/jit/attributes.h]: This stuff duplicates code #97 so I would rather not merge this part.

File: torch/csrc/jit/ir.h [hunk 3]
```diff
@@ -707,7 +707,13 @@ friend struct Node;
     n->i_(kDim, dim);
     return n;
   }
+  Node * createUndefined() {
```
Comment (zdevito) [inline torch/csrc/jit/ir.h:709]: Why not just kUndefined? The checks inside convolution.cpp and batchnorm.cpp would be simpler.

**Review by zdevito — APPROVED** [review]
Looks good, but I don't want to merge the attribute output because it conflicts with the other pr. The other parts are good to merge.

File: torch/csrc/jit/ir.h (hunk 3 above)
Comment (ezyang) [reply torch/csrc/jit/ir.h:709]: It seemed more parsimonious. If we add a new kUndefined field, what is the difference between a kConstant with an undefined tensor, versus a kUndefined? Perhaps you are suggesting we should apply the invariant that kConstant is always defined.

File: torch/csrc/jit/attributes.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/jit/attributes.h]: easy enough :)

File: torch/csrc/autograd/functions/special.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/special.cpp]: I think this code will last pretty long. In principle we can get rid of it once we stop allowing null variables, but that won't be happening for this release.

File: test/common.py [hunk 4]
```diff
@@ -296,10 +297,13 @@ def remove_prefix(text, prefix):
             try:
                 with open(expected_file) as f:
                     expected = f.read()
-            except FileNotFoundError:
-                raise RuntimeError(
-                    ("No expect file exists; to accept the current output, run:\n"
-                     "python {} {} --accept").format(__main__.__file__, munged_id))
+            except EnvironmentError as e:
```
Comment (apaszke) [inline test/common.py]: Why did you change it? Are you sure this works for both py2 and py3?

File: test/expect/TestModels.test_alexnet-pbtxt.expect [hunk 5]
```diff
@@ -304,12 +304,12 @@ node {
...
-    name: "ratio"
-    f: 0.5
+    name: "is_test"
+    i: 0
   }
   attribute {
-    name: "is_test"
-    f: 0
+    name: "ratio"
+    f: 0.5
```
Comment (apaszke) [inline test/expect/TestModels.test_alexnet-pbtxt.expect]: Is protobuf attr ordering deterministic?

File: test/expect/TestModels.test_densenet-121.expect [hunk 6]
```diff
@@ -1,4 +1,488 @@
...
+graph(%1 : Double(64, 3, 7, 7)
```
Comment (apaszke) [inline test/expect/TestModels.test_densenet-121.expect:1]: I thought we discussed expects for large models. Having the same graph pattern for 100000 layers doesn't really exercise any of our code and changes are not interpretable

File: test/test_jit.py [hunk 7]
```diff
@@ -226,6 +233,42 @@ def test_backward(self):
...
+        self.assertExpected(str(trace))
+        # NB: Purposely NOT testing protobuf export here
+
+    @skipIfNoTorchVision
+    def test_densenet(self):
+        x = Variable(torch.randn(10, 3, 224, 224).fill_(1.0), requires_grad=True)
+        dense121 = torchvision.models.DenseNet(num_init_features=64, growth_rate=32,
+                                               block_config=(6, 12, 24, 16))
+        trace, _ = torch.jit.record_trace(dense121, x)
+        # Densenet trace is pretty large, so we don't assert on it
```
Comment (apaszke) [inline test/test_jit.py:271]: ?? But we still have an expect file? Maybe just use a smaller model?

File: torch/csrc/autograd/functions/special.cpp [hunk 8]
```diff
@@ -38,9 +40,10 @@ auto Eval::getSubgraph(const variable_list& inputs, const variable_list& outputs
   }
 
   // Regular DFS data structures
-  std::unordered_set<Function*> seen;
+  std::unordered_set<Function*> seen { nullptr };
   std::vector<Function*> queue;
   for (auto & output : outputs) {
+    if (!output) continue;
```
Comment (apaszke) [inline torch/csrc/autograd/functions/special.cpp]: There's an invariant that outputs in this function are relevant, and nullptr outputs are not. Remove this line

File: torch/csrc/autograd/functions/special.cpp [hunk 9]
```diff
@@ -104,11 +111,13 @@ variable_list Eval::filterRelevantOutputs(const variable_list& inputs, const var
 auto Eval::computeInputOrder(const variable_list& inputs, const placeholder_list& inherited_placeholders) -> edge_order {
   edge_order input_order;
   int idx = 0;
-  for (auto & input : inputs)
+  for (auto & input : inputs) {
+    if (!input) continue;
```
Comment (apaszke) [inline torch/csrc/autograd/functions/special.cpp]: Can you mark all places that will have to be removed with some kind of a TODO?

File: test/common.py (hunk 4 above)
Comment (ezyang) [reply test/common.py]: Not tested, but apparently EnvironmentError == IOError in Python 3.3 and FileNotFound error is a subclass of IOError. https://docs.python.org/3/library/exceptions.html

File: test/expect/TestModels.test_alexnet-pbtxt.expect (hunk 5 above)
Comment (ezyang) [reply test/expect/TestModels.test_alexnet-pbtxt.expect]: I spent a while yesterday trying to figure out if this is the case. I think the answer is, no, it's not, and so we will have to write our own protobuf diff'er.

File: test/expect/TestModels.test_densenet-121.expect (hunk 6 above)
Comment (ezyang) [reply test/expect/TestModels.test_densenet-121.expect:1]: Actually, test_jit.py is not testing on this expect; I just forgot to remove it from the commit ^^"

File: test/test_jit.py (hunk 7 above)
Comment (ezyang) [reply test/test_jit.py:271]: Yes I need to delete the expect file

File: test/common.py (hunk 4 above)
Comment (ezyang) [reply test/common.py]: (Oh, I changed it because the old code doesn't work with Python 2)

File: test/expect/TestModels.test_densenet-121.expect (hunk 6 above)
Comment (ezyang) [reply test/expect/TestModels.test_densenet-121.expect:1]: No, in fact, these are different expect files (this is a models test, which is going to be terrible until we fix tests.)

File: torch/csrc/autograd/functions/special.cpp (hunk 9 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/special.cpp]: Done

File: torch/csrc/autograd/functions/special.cpp (hunk 8 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/special.cpp]: Done

File: torch/csrc/jit/attributes.h (hunk 2 above)
Comment (ezyang) [reply torch/csrc/jit/attributes.h]: Done

**Comment by ezyang** [comment]
merged.
